clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -24,123 +24,380 @@ package common
+		}
+	}
+}
+
+func (w *waitForControlledPodsRunningMeasurement) createObjectLocked(c clientset.Interface, obj runtime.Object) error {
+	key, err := runtimeobjects.CreateMetaNamespaceKey(obj)
 	if err != nil {
-		return summaries, err
+		return fmt.Errorf("meta key creation error: %v", err)
 	}
-	fieldSelector, err := util.GetStringOrDefault(config.Params, "fieldSelector", "")
+	checker, err := w.waitForRuntimeObject(c, obj)
 	if err != nil {
-		return summaries, err
+		return fmt.Errorf("waiting for %v error: %v", key, err)
 	}
-	timeout, err := util.GetDurationOrDefault(config.Params, "timeout", defaultWaitForRCPodsTimeout)
+	time.AfterFunc(w.operationTimeout, func() {
+		checker.terminate()
+	})
+	w.checkerMap[key] = checker
+	return nil
+}
+
+func (w *waitForControlledPodsRunningMeasurement) deleteObjectLocked(c clientset.Interface, obj runtime.Object) error {
+	key, err := runtimeobjects.CreateMetaNamespaceKey(obj)
 	if err != nil {
-		return summaries, err
+		return fmt.Errorf("meta key creation error: %v", err)
+	}
+	if checker, exists := w.checkerMap[key]; exists {
+		checker.terminate()
+		delete(w.checkerMap, key)
 	}
+	return nil
+}
 
-	runtimeObjectsList, err := runtimeobjects.ListRuntimeObjectsForKind(config.ClientSet, kind, namespace, labelSelector, fieldSelector)
+func (w *waitForControlledPodsRunningMeasurement) updateOpResourceVersion(runtimeObj runtime.Object) error {
+	if runtimeObj == nil {
+		return nil
+	}
+	version, err := runtimeobjects.GetResourceVersionFromRuntimeObject(runtimeObj)
 	if err != nil {
-		return summaries, err
+		return fmt.Errorf("retriving resource version error: %v", err)
 	}
-	var wg wait.Group
-	errList := util.NewErrorList()
-	var runningRuntimeObjects int32
-	for objectCounter := 0; objectCounter < len(runtimeObjectsList); objectCounter++ {
-		objectIndex := objectCounter
-		wg.Start(func() {
-			if err := waitForRuntimeObject(config.ClientSet, runtimeObjectsList[objectIndex], timeout); err != nil {
-				errList.Append(fmt.Errorf("waiting for %v error: %v", kind, err))
-				return
-			}
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	if version > w.opResourceVersion {
+		w.opResourceVersion = version
+	}
+	return nil
+}
 
-			atomic.AddInt32(&runningRuntimeObjects, 1)
-			objName, err := runtimeobjects.GetNameFromRuntimeObject(runtimeObjectsList[objectIndex])
-			if err != nil {
-				errList.Append(fmt.Errorf("reading object name error: %v", err))
-				return
-			}
-			objNamespace, err := runtimeobjects.GetNamespaceFromRuntimeObject(runtimeObjectsList[objectIndex])
-			if err != nil {
-				errList.Append(fmt.Errorf("reading object namespace error: %v", err))
-				return
-			}
-			glog.Infof("%s: %s (%s) has all pods running", waitForControlledPodsRunningName, objName, objNamespace)
-		})
+func (w *waitForControlledPodsRunningMeasurement) getObjectCountAndMaxVersion(c clientset.Interface) (int, uint64, error) {
+	var desiredCount int
+	var maxResourceVersion uint64
```

Comment:
Please add a comment why the approach with count + resourceVersion works (something that we discussed offline).

Reply:
Done.